pkg/tools/prefix_pool/prefixpool.go

```diff
@@ -0,0 +1,642 @@
+					changes++
+				}
+			}
+			leftPrefixes := []*net.IPNet{}
+			// Put all not merged values
+			for _, value := range base {
+				leftPrefixes = append(leftPrefixes, value)
+			}
+			newPrefixByPrefixLen[basePrefix] = leftPrefixes
+			newPrefixByPrefixLen[basePrefix-1] = append(newPrefixByPrefixLen[basePrefix-1], parentPrefix...)
+		}
+		if changes == 0 {
+			// All is merged, we could exit
+			result = []string{}
+			for _, value := range newPrefixByPrefixLen {
+				for _, v := range value {
+					result = append(result, v.String())
+				}
+			}
+			return result, nil
+		}
+		prefixByPrefixLen = newPrefixByPrefixLen
+	}
+}
+
+func subnet(ipnet *net.IPNet, subnet_index int) (*net.IPNet, error) {
+	mask := ipnet.Mask
+
+	parentLen, addrLen := mask.Size()
+	newPrefixLen := parentLen + 1
+	if newPrefixLen > addrLen {
+		return nil, errors.Errorf("insufficient address space to extend prefix of %d", parentLen)
+	}
+
+	if uint64(subnet_index) > 2 {
+		return nil, errors.Errorf("prefix extension does not accommodate a subnet numbered %d", subnet_index)
+	}
+
+	return &net.IPNet{
+		IP:   setNetIndexInIP(ipnet.IP, subnet_index, newPrefixLen),
+		Mask: net.CIDRMask(newPrefixLen, addrLen),
+	}, nil
+}
+
+func AddressCount(prefixes ...string) uint64 {
+	var c uint64 = 0
+	for _, pr := range prefixes {
+		c += addressCount(pr)
+	}
+	return c
+}
+
+func addressCount(pr string) uint64 {
+	_, network, _ := net.ParseCIDR(pr)
+	prefixLen, bits := network.Mask.Size()
+	return 1 << (uint64(bits) - uint64(prefixLen))
+}
+
+func setNetIndexInIP(ip net.IP, num int, prefixLen int) net.IP {
+	ipInt, totalBits := fromIP(ip)
+	bigNum := big.NewInt(int64(num))
+	bigNum.Lsh(bigNum, uint(totalBits-prefixLen))
+	ipInt.Or(ipInt, bigNum)
+	return toIP(ipInt, totalBits)
+}
+
+func clearNetIndexInIP(ip net.IP, prefixLen int) net.IP {
+	ipInt, totalBits := fromIP(ip)
+	ipInt.SetBit(ipInt, totalBits-prefixLen, 0)
+	return toIP(ipInt, totalBits)
+}
+
+func toIP(ipInt *big.Int, bits int) net.IP {
+	ipBytes := ipInt.Bytes()
+	ret := make([]byte, bits/8)
+	// Pack our IP bytes into the end of the return array,
+	// since big.Int.Bytes() removes front zero padding.
+	for i := 1; i <= len(ipBytes); i++ {
+		ret[len(ret)-i] = ipBytes[len(ipBytes)-i]
+	}
+	return net.IP(ret)
+}
+
+func fromIP(ip net.IP) (*big.Int, int) {
+	val := &big.Int{}
+	val.SetBytes([]byte(ip))
+	i := len(ip)
+	if i == net.IPv4len {
+		return val, 32
+	} // else if i == net.IPv6len
+	return val, 128
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func MaxCommonPrefixSubnet(s1, s2 *net.IPNet) *net.IPNet {
+	rawIp1, n1 := fromIP(s1.IP)
+	rawIp2, _ := fromIP(s2.IP)
+
+	xored := &big.Int{}
+	xored.Xor(rawIp1, rawIp2)
+	maskSize := leadingZeros(xored, n1)
+
+	m1, bits := s1.Mask.Size()
+	m2, _ := s2.Mask.Size()
+
+	mask := net.CIDRMask(min(min(m1, m2), maskSize), bits)
+	return &net.IPNet{
+		IP:   s1.IP.Mask(mask),
+		Mask: mask,
+	}
+}
+
+func IpToNet(ipAddr net.IP) *net.IPNet {
```

Comment:
Can we make this thing private?

Reply:
It is also used in `prefixcollector` package, it's better to leave it public

Reply:
Deleted this function also, because it's used in `prefixcollector` only